File: frontend/src/api/announcement.ts
import api from './client'

export interface Announcement {
  id: number
  title: string
  content: string
  author_id: number
  is_pinned: boolean
  is_published: boolean
  published_at: string | null
  created_at: string
  updated_at: string
}

export interface PaginatedAnnouncements {
  items: Announcement[]
  total: number
  page: number
  page_size: number
}

export const announcementApi = {
  listPublished: (page = 1, pageSize = 10) =>
    api.get<PaginatedAnnouncements>('/announcements', { params: { page, page_size: pageSize } }),

  getById: (id: number) =>
    api.get<Announcement>(`/announcements/${id}`),

  create: (data: { title: string; content: string; is_pinned: boolean }) =>
    api.post<Announcement>('/announcements', data),

  update: (id: number, data: Partial<{ title: string; content: string; is_pinned: boolean; is_published: boolean }>) =>
    api.put<Announcement>(`/announcements/${id}`, data),

  delete: (id: number) =>
    api.delete(`/announcements/${id}`),
}

File: backend/app/schemas/announcement.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class AnnouncementCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    content: str = Field(..., min_length=1)
    is_pinned: bool = False


class AnnouncementUpdate(BaseModel):
    title: str | None = Field(None, max_length=200)
    content: str | None = None
    is_pinned: bool | None = None
    is_published: bool | None = None


class AnnouncementOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    content: str
    author_id: int
    is_pinned: bool
    is_published: bool
    published_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class PaginatedAnnouncements(BaseModel):
    items: list[AnnouncementOut]
    total: int
    page: int
    page_size: int
